src/lib/comms/sol-bluetooth-impl-linux.c

```diff
@@ -0,0 +1,683 @@
+
+    r = sol_bus_map_cached_properties(ctx->bluez, path,
+        "org.bluez.Adapter1",
+        adapter_properties,
+        adapter_property_changed, ctx);
+    SOL_INT_CHECK_GOTO(r, < 0, error_map);
+
+    if (ctx->enabled_callback) {
+        r = adapter_set_powered(ctx, ctx->adapter_path, true);
+        SOL_INT_CHECK(r, < 0);
+    }
+
+    return;
+
+error_map:
+    free(ctx->adapter_path);
+    ctx->adapter_path = NULL;
+}
+
+
+static void
+destroy_device(struct device_info *device)
+{
+    struct sol_bt_device_info *info;
+
+    info = &device->info;
+    free(info->name);
+    sol_vector_clear(&info->uuids);
+
+    free(device->path);
+    free(device);
+}
+
+static struct device_info *find_device_by_path(struct context *ctx, const char *path)
+{
+    struct device_info *d;
+    uint16_t i;
+
+    SOL_NULL_CHECK(path, NULL);
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&ctx->devices, d, i) {
+        if (streq(d->path, path))
+            return d;
+    }
+    return NULL;
+}
+
+static void
+adapter_removed(void *data, const char *path)
+{
+
+}
+
+enum {
+    DEVICE_PROPERTY_ADDRESS = 0,
+    DEVICE_PROPERTY_NAME,
+    DEVICE_PROPERTY_UUIDS,
+    DEVICE_PROPERTY_RSSI,
+};
+
+static bool
+device_property_address_set(void *data, const char *path, sd_bus_message *m)
+{
+    struct context *ctx = data;
+    struct device_info *d;
+    struct sol_bt_device_info *info;
+    const char *address;
+    int r;
+
+    d = find_device_by_path(ctx, path);
+    SOL_NULL_CHECK_GOTO(d, skip);
+
+    info = &d->info;
+    if (info->addr.family)
+        goto skip;
+
+    r = sd_bus_message_read_basic(m, 's', &address);
+    SOL_INT_CHECK_GOTO(r, < 0, skip);
+
+    sol_network_link_addr_from_str(&info->addr, address);
+
+    return true;
+
+skip:
+    r = sd_bus_message_skip(m, "s");
+    SOL_INT_CHECK(r, < 0, false);
+
+    return false;
+}
+
+static bool
+device_property_name_set(void *data, const char *path, sd_bus_message *m)
+{
+    struct context *ctx = data;
+    struct device_info *d;
+    struct sol_bt_device_info *info;
+    const char *name;
+    int r;
+
+    d = find_device_by_path(ctx, path);
+    SOL_NULL_CHECK_GOTO(d, skip);
+
+    info = &d->info;
+
+    if (info->name)
+        goto skip;
+
+    r = sd_bus_message_read_basic(m, 's', &name);
+    SOL_INT_CHECK_GOTO(r, < 0, skip);
+
+    /* When we receive the property again we will try to allocate again */
+    info->name = strdup(name);
+
+    return true;
+
+skip:
+    r = sd_bus_message_skip(m, "s");
+    SOL_INT_CHECK(r, < 0, false);
+
+    return false;
+}
+
+static bool
+device_property_uuids_set(void *data, const char *path, sd_bus_message *m)
+{
+    struct context *ctx = data;
+    struct device_info *d;
+    struct sol_bt_device_info *info;
+    struct sol_bt_uuid *u = NULL;
+    char **uuids = NULL, **uuid;
+    int r;
+
+    d = find_device_by_path(ctx, path);
+    SOL_NULL_CHECK_GOTO(d, skip);
+
+    info = &d->info;
+
+    r = sd_bus_message_read_strv(m, &uuids);
```

Comment:
can't you open the array, iterate using on-wire data, then avoid the strv memory allocation and free? you just need it to decode the uuid to our sol_bt_uuid